### PR TITLE
5 rk45 jax solver returning nan solution

### PR DESCRIPTION
**Pull request by jamie-cunningham**
This change moves JAX functions to a separate module and allows toggling use of JAX via an environment variable.

### CODE REVIEW

**Comment by jamie-cunningham** [comment]
@keatincf I moved the JAX functionality to its own package and added an environment variable check for enabling JAX. Unfortunately using JAX seems to add significant overhead, resulting in a 10x slowdown vs numpy. I couldn't nail down why during profiling.

I also haven't directly addressed the NaN issue yet. I couldn't find any good guidance in the solver docs for how to deal with invalid solutions. Note that I used the [`diffrax`](https://docs.kidger.site/diffrax/) package for JAX solvers as it appears the JAX ode solver has been declared out of scope by the JAX team and no longer is actively developed. Any thoughts are welcome.

**Comment by keatincf** [comment]
test_ode is failing because the _clip_state_dot_direct was moved out of ODEDynamics.

I spent a little bit of time trying to see if there was anything I could see that could be causing the slowdown due to JAX.  I didn't see anything that would make me think JAX was recompling more than expected.  Removing the jit decorators causes a further slow down, but only a second or two extra slowdown.  I was worried that maybe the functions being compiled weren't pure enough be compiled properly.

I also tried to utilize a GPU enabled JAX, but that generated a larger slowdown.  I'm not sure if that was partially caused by a driver/CUDA version mismatch (Ubuntu's repos have CUDA 12.4 and JAX is using CUDA 12.6).

It looks like the biggest slowdown might be using diffrax compared to scipy.integrate.

My guess is that JAX does not provide any major improvement on CPUs compared to numpy/scipy.  It could provide better performance on a GPU/TPU, but we may need to adjust the code more to properly take advantage of JAX.

It should be ok to keep JAX in since you've noted that the use of JAX is experimental.

**Comment by jamie-cunningham** [comment]
> test_ode is failing because the _clip_state_dot_direct was moved out of ODEDynamics.
> 
> I spent a little bit of time trying to see if there was anything I could see that could be causing the slowdown due to JAX. I didn't see anything that would make me think JAX was recompling more than expected. Removing the jit decorators causes a further slow down, but only a second or two extra slowdown. I was worried that maybe the functions being compiled weren't pure enough be compiled properly.
> 
> I also tried to utilize a GPU enabled JAX, but that generated a larger slowdown. I'm not sure if that was partially caused by a driver/CUDA version mismatch (Ubuntu's repos have CUDA 12.4 and JAX is using CUDA 12.6).
> 
> It looks like the biggest slowdown might be using diffrax compared to scipy.integrate.
> 
> My guess is that JAX does not provide any major improvement on CPUs compared to numpy/scipy. It could provide better performance on a GPU/TPU, but we may need to adjust the code more to properly take advantage of JAX.
> 
> It should be ok to keep JAX in since you've noted that the use of JAX is experimental.

Yeah I also tried just using `scipy.integrate` in a JIT wrapper but there were numpy vs jax issues that led me to diffrax. It doesn't seem like JAX handles third-party packages using numpy very well.

**Comment by keatincf** [comment]
Kyle brought up a use case that this solution doesn't cover, or makes it harder to achieve.  When running the inspection environment, you may only want JAX to be used for the dynamcis solver used by RTA and not use JAX when stepping the dynamics of the entities.  Using the environment variable would make it harder, if not impossible, to have multiple solvers in an instance, with some using JAX and some not.

**Comment by jamie-cunningham** [comment]
JAX flags have been moved to the class level. Class objects can now decide if they want to use JAX or not. JAX will only be used if it is available in the environment.

File: safe_autonomy_simulation/sims/spacecraft/point_model.py [hunk 1]
```diff
@@ -189,6 +191,7 @@ def __init__(
         state_dot_min: typing.Union[float, np.ndarray] = -np.inf,
         state_dot_max: typing.Union[float, np.ndarray] = np.inf,
         integration_method: str = "RK45",
+        use_jax: bool = False
```
Comment (keatincf) [inline safe_autonomy_simulation/sims/spacecraft/point_model.py:194]: The dynamics can specify whether jax should be used, but the models using them don't have a way to pass that down.  The dynamics will always be forced to not use jax.

**Review by keatincf — CHANGES_REQUESTED** [review]
I can't tell if the issue I'm having is Python's loose typing or the intermingling of jax and non-jax code.  It's hard to track whether types are staying consistent and whether casting is happening too often.

**Comment by jamie-cunningham** [comment]
> casting

Yeah Jax requires that input arrays to JIT functions are Jax types hence the casting. Since we aren't globally setting a Jax flag then it doesn't make sense to have Jax arrays in non-Jax compatible code. When there was a global flag we could just make everything a Jax type internally. Now we need to make sure that inputs to JIT functions are Jax types.

I can go back in and see if I can beef up the type hinting. Unfortunately checking a Jax type requires Jax to be available at runtime. We could make Jax a non-optional dependency again since we are being explicit about its use.

File: safe_autonomy_simulation/dynamics/dynamics.py [hunk 2]
```diff
@@ -56,12 +64,22 @@ def step(
         Tuple[np.ndarray, np.ndarray]
             Tuple of the system's next state and the state's instantaneous time derivative at the end of the step
         """
+        state = cast_jax(state, use_jax=self.use_jax)
+        control = cast_jax(control, use_jax=self.use_jax)
         next_state, state_dot = self._step(step_size, state, control)
-        next_state = self.np.clip(next_state, self.state_min, self.state_max)
+
+        next_state = self.np.clip(
```
Comment (keatincf) [inline safe_autonomy_simulation/dynamics/dynamics.py:71]: I was playing around with jax and it looks like the jnp `clip` is slower than the numpy version, even if you use jit.  So, we can probably just use `numpy.clip` all the time.  Not sure if this provides a simplification that will allow us to localize conversion between jax and numpy to a smaller area in the code base

File: safe_autonomy_simulation/opts.py [hunk 3]
```diff
@@ -3,17 +3,19 @@
 import os
 
 
-def use_jax():
-    """Whether or not to use JAX for numerical computing and JIT compilation
+def jax_available():
```
[the remark this answers is not in the document]
Comment (jamie-cunningham) [reply safe_autonomy_simulation/opts.py]: The latest commit makes jax required and gets rid of this function completely.

File: safe_autonomy_simulation/utils/conversion.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import numpy as np
+import safe_autonomy_simulation
+
+if safe_autonomy_simulation.jax_available():
+    import jax.numpy as jnp
+
+
+def cast_jax(arr: np.ndarray, use_jax: bool) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (jamie-cunningham) [reply safe_autonomy_simulation/utils/conversion.py]: Makes sense to me. Updated in the latest commit so that the caller manually casts to jax. This function has been removed.

File: safe_autonomy_simulation/dynamics/dynamics.py [hunk 5]
```diff
@@ -57,12 +64,22 @@ def step(
         Tuple[np.ndarray, np.ndarray]
             Tuple of the system's next state and the state's instantaneous time derivative at the end of the step
         """
+        state = cast_jax(state, use_jax=self.use_jax)
+        control = cast_jax(control, use_jax=self.use_jax)
         next_state, state_dot = self._step(step_size, state, control)
```
[the remark this answers is not in the document]
Comment (jamie-cunningham) [reply safe_autonomy_simulation/dynamics/dynamics.py:63]: Method signatures have been reviewed and updated in the latest commit. Methods and functions which might be called inside a JIT function have had their signatures updated to allow jax arrays. This is required for jax JIT functions to work. Arrays are now only cast to jax arrays before calling JIT functions.

File: safe_autonomy_simulation/dynamics/dynamics.py (hunk 2 above)
Comment (jamie-cunningham) [reply safe_autonomy_simulation/dynamics/dynamics.py]: If its faster than we can just use numpy here! The state limits might be jax arrays but it seems numpy can handle mixing types just fine. I've committed a change to just use `np.clip` here.

File: safe_autonomy_simulation/sims/spacecraft/point_model.py (hunk 1 above)
Comment (jamie-cunningham) [reply safe_autonomy_simulation/sims/spacecraft/point_model.py:198]: I've exposed jax to the `safe_autonomy_simulation.sims` entities and the point mass integrators

**Review by keatincf — COMMENTED** [review]
Have some new comments.  We could potentially move some of the comments to other issues, if we find that is appropriate.

File: safe_autonomy_simulation/dynamics/ode.py [hunk 6]
```diff
@@ -93,143 +97,88 @@ def compute_state_dot(
...
+        # Clip state derivative values
+        state_dot = clip_fn(
+            state_dot=state_dot,
+            s_min=self.state_dot_min,
+            s_max=self.state_dot_max,
         )
-        state_dot = self.np.where(
-            upper_bounded_states, upper_bounded_clipped, state_dot
+        # Clip state derivative values to ensure state remains within bounds
+        state_dot = clip_at_state_limits_fn(
```
[the remark this answers is not in the document]
Comment (jamie-cunningham) [reply safe_autonomy_simulation/dynamics/ode.py:132]: So this function isn't actually clipping state_dot to the state limits, its doing what you said. The first clip clips state_dot to the state_dot limits, the second clip clips state_dot such that it doesn't push state beyond the state limits. We could combine these.

[the remark this answers is not in the document]
Comment (keatincf) [reply safe_autonomy_simulation/dynamics/ode.py:132]: I guess I'm not seeing how the clipping is producing a state_dot that would keep the state within limits.  We're applying state limits to the state_dot, but that could still produce an out of bounds state.  For example, if I have a 1-D position currently at 9 with limits [-10,10] and a velocity with no limits, having a velocity that clips to the max state limit (10) would produce a new position of 19, which would be out of bounds.

[the remark this answers is not in the document]
Comment (keatincf) [reply safe_autonomy_simulation/dynamics/ode.py:124]: That makes sense.  No change needed then.

File: safe_autonomy_simulation/dynamics/ode.py [hunk 7]
```diff
@@ -93,143 +97,88 @@ def compute_state_dot(
...
+            )
+            state = jnp.array(state)
+            control = jnp.array(control)
+
+        next_state, state_dot, self.trajectory, self.trajectory_t = step_fn(
+            f=self.compute_state_dot,
+            step_size=step_size,
+            state=state,
+            control=control,
+        )
```
[the remark this answers is not in the document]
Comment (keatincf) [reply safe_autonomy_simulation/dynamics/ode.py:167]: In the case where `use_jax` is true, we could convert the jnp array back to a np array after the JIT function call so that jnp isn't being used outside of the JIT function.  My worry would be that we start with a numpy array, convert to jax numpy, and then the state just stays as a jax numpy array the rest of the time.  It seems like jax numpy is slower than numpy if you're not using JIT or offloading to some accelerator.  If you find that converting back to numpy doesn't really overcome the initial conversion to jax numpy, I'm good with leaving it out.

[the remark this answers is not in the document]
Comment (jamie-cunningham) [reply safe_autonomy_simulation/dynamics/ode.py:167]: I think it's something we should investigate in a separate issue. It's a good point.

[the remark this answers is not in the document]
Comment (keatincf) [reply safe_autonomy_simulation/dynamics/ode.py:167]: Sounds good.  I'll open a new issue.  Marking this resolved.

**Review by keatincf — APPROVED** [review]
